File: m7-lesson3-start/um-webapp/src/main/java/com/baeldung/um/service/impl/PrivilegeServiceImpl.java
package com.baeldung.um.service.impl;

import org.apache.commons.lang3.tuple.Triple;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.baeldung.common.persistence.service.AbstractService;
import com.baeldung.common.search.ClientOperation;
import com.baeldung.um.persistence.dao.IPrivilegeJpaDao;
import com.baeldung.um.persistence.model.Privilege;
import com.baeldung.um.persistence.util.SearchUtilSec;
import com.baeldung.um.service.IPrivilegeService;

@Service
@Transactional
public class PrivilegeServiceImpl extends AbstractService<Privilege>implements IPrivilegeService {

    @Autowired
    IPrivilegeJpaDao dao;

    public PrivilegeServiceImpl() {
        super(Privilege.class);
    }

    // API

    // find

    @Override
    public Privilege findByName(final String name) {
        return getDao().findByName(name);
    }

    // Spring

    @Override
    protected final IPrivilegeJpaDao getDao() {
        return dao;
    }

    @Override
    public Specification<Privilege> resolveConstraint(final Triple<String, ClientOperation, String> constraint) {
        return SearchUtilSec.resolveConstraint(constraint, Privilege.class);
    }

    @Override
    protected JpaSpecificationExecutor<Privilege> getSpecificationExecutor() {
        return dao;
    }

}

File: m12-lesson2/um-webapp/src/main/kotlin/com/baeldung/um/service/impl/PrivilegeServiceImpl.kt
package com.baeldung.um.service.impl

import com.baeldung.common.persistence.service.AbstractService
import com.baeldung.um.persistence.dao.IPrivilegeJpaDao
import com.baeldung.um.persistence.model.Privilege
import com.baeldung.um.service.IPrivilegeService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class PrivilegeServiceImpl(private val dao: IPrivilegeJpaDao) : AbstractService<Privilege>(), IPrivilegeService {

    // API

    // find

    override fun findByName(name: String): Privilege? = getDao().findByName(name)

    // Spring

    override fun getDao() = dao

    override fun getSpecificationExecutor() = dao
}
